In [1]:
# experimental: you can run this cell to see your code full-width
from IPython.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

In [2]:
# cell for pair-programming magic link

#  run this cell without changing it. It only works on Noteable.
! printf "\n\nLink below can be send to someone to share your Noteable session:\n\nhttps://noteable.edina.ac.uk/user/$(jupyter notebook list | grep -oP '(?<=user\/).*(?=\/\?)' )/tree?token=$( jupyter notebook list | grep -oP '(?<=token=).*(?= ::)' )\n\n"



Link below can be send to someone to share your Noteable session:

https://noteable.edina.ac.uk/user/d7i6m6qmi3q5tglmiqr86m/tree?token=OyaJoCPcRPevE70HdUwVrg



# Who survived on Titanic?

In this lab we will use a famous leanring dataset about passangers of titanic. Titanic was a big passager boat which sunk on 15 April 1912. Lots of people died, but we have quite detailed information about passangers and whether they survived. On this lab we will investigate relationship between those variables (eg. wealth, age, gender, cabin number) and survival.

In [1]:
import pandas as pd
import numpy as np

In [33]:
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
# this version of titanic dataset is taken from Kaggle
# https://www.kaggle.com/competitions/titanic/data?select=test.csv
passengers

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [4]:
set(passengers['SibSp'])

{0, 1, 2, 3, 4, 5, 8}

In [5]:
set(passengers['Parch'])

{0, 1, 2, 3, 4, 5, 6}

In [6]:
set(passengers['Embarked'])

{<NA>, 'C', 'Q', 'S'}

In [3]:
passengers = pd.read_csv('./data/titanic_kaggle_train.csv').convert_dtypes()
passengers

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  <NA>      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171     7.25  <NA>        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282    7.925  <NA>        S  
3        0            113803     53.1  C123        S  
4        0            373450     8.05  <NA>        S  
..     ...               ...      ...   ...      ...  
886      0            211536     13.0  <NA>        S  
887      0            112053     30.0   B42        S  
888      2        W./C. 6607    23.45  <NA>        S  
889      0            111369     30.0  C148        C  
890      0            370376     7.75  <NA>        Q  

[891 rows x 12 columns]

# Task 1: Descriptive statistics

Describe the dataset in a set of bulletpoints (use Pandas to give you the answers). Try to backup each bulletpoint with separate pieces of python code.

Below are just some examples to get you started, but explore numbers and types of things (rows, columns). What are max min and mean of the numeric ones? Which columns have missing values and how many? etc.

For exaplanations of what columns mean see https://www.kaggle.com/competitions/titanic/data?select=test.csv

### A. What columns and of what type did you find? How many rows and columns?

In [7]:
passengers.shape

(891, 12)

In [12]:
passengers.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [13]:
type(passengers.columns)

pandas.core.indexes.base.Index

In [14]:
type(passengers)

pandas.core.frame.DataFrame

In [18]:
passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [61]:
passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    Int64   
 1   Survived     891 non-null    bool    
 2   Pclass       891 non-null    category
 3   Name         891 non-null    string  
 4   Sex          891 non-null    string  
 5   Age          714 non-null    Float64 
 6   SibSp        891 non-null    category
 7   Parch        891 non-null    category
 8   Ticket       891 non-null    string  
 9   Fare         891 non-null    Float64 
 10  Cabin        204 non-null    string  
 11  Embarked     889 non-null    category
dtypes: Float64(2), Int64(1), bool(1), category(4), string(4)
memory usage: 56.8 KB


In [122]:
passengers['Survived'] = passengers['Survived'].astype('bool')
passengers['Pclass'] = passengers['Pclass'].astype('category')
passengers['SibSp'] = passengers['SibSp'].astype('category')
passengers['Parch'] = passengers['Parch'].astype('category')
passengers['Embarked'] = passengers['Embarked'].astype('category')

passengers['Embarked'] = passengers['Embarked'].cat.rename_categories({'C': 'Cherbourg', 'Q': 'Queenstown', 'S': 'Southampton'})
print('Embarked:', passengers['Embarked'].cat.categories)

passengers['Pclass'] = passengers['Pclass'].cat.rename_categories({1: '1st', 2: '2nd', 3: '3rd'})
print('Pclass:', passengers['Pclass'].cat.categories)

print(passengers.info())

Embarked: Index(['Cherbourg', 'Queenstown', 'Southampton'], dtype='object')
Pclass: Index(['1st', '2nd', '3rd'], dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    Int64   
 1   Survived     891 non-null    bool    
 2   Pclass       891 non-null    category
 3   Name         891 non-null    string  
 4   Sex          891 non-null    string  
 5   Age          714 non-null    Float64 
 6   SibSp        891 non-null    category
 7   Parch        891 non-null    category
 8   Ticket       891 non-null    string  
 9   Fare         891 non-null    Float64 
 10  Cabin        204 non-null    string  
 11  Embarked     889 non-null    category
dtypes: Float64(2), Int64(1), bool(1), category(4), string(4)
memory usage: 56.8 KB
None


In [87]:
print("Mean:", passengers.Age.mean())
# ...
# ...
# (you can add more cells wirh + button in the top menu)

Mean: 29.69911764705882


**Answer A:**

there are 12 columns (+ index) and 891 rows. 

Columns have names: 'PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'

Name and Sex columns are an object (likely a string), so we converted them to `strings`. Age and Fare are Floats, and Sibling/Parent rows are integers.

Age had a mean 29 years, and max and min are ... [you continue]

# Task 2. Answers basic questions: Select some rows, some columns and run some simple statistics

What makes a good **'Basic Question'**? Question which you can explore by selecting only some rows and columns.

Example Basic Questions:

- just select name and age components of the data frame. Sort them by Age and get the names of 10 oldest passangers
- Pick the fares of pasangers who survived and did not survive and compare their ticket class (Pclass), Fare cost and sex. Is there a large difference in the means?

**Itentify 5 Basic Questions** like above and answer them with a simple sentence. Provide python code which will find you that answer. 

Call them 'A', 'B', 'C', ...

### To get you started, here is some starting code. Read it, make sure you understand it. Feel free to modify it.

How to select only some rows and some columns?

Note about referencing: iloc[rows, columns] is used when referencing with numbers, and loc[rows, columns] is used when referencing by names. 

Also colon `:` means `'all of them'`

In [60]:
# first passanger, their data in column indexed 4
passengers.iloc[0,4]

'male'

In [65]:
passengers

PassengerId  Survived Pclass  \
0              1     False    3rd   
1              2      True    1st   
2              3      True    3rd   
3              4      True    1st   
4              5     False    3rd   
..           ...       ...    ...   
886          887     False    2nd   
887          888      True    1st   
888          889     False    3rd   
889          890      True    1st   
890          891     False    3rd   

                                                  Name     Sex   Age SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0     1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0     1   
2                               Heikkinen, Miss. Laina  female  26.0     0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0     1   
4                             Allen, Mr. William Henry    male  35.0     0   
..                                                 ...     ...   ...   ...   
886                              Montvila, Rev. Juozas    male  27.0     0   
887                       Graham, Miss. Margaret Edith  female  19.0     0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  <NA>     1   
889                              Behr, Mr. Karl Howell    male  26.0     0   
890                                Dooley, Mr. Patrick    male  32.0     0   

    Parch            Ticket     Fare Cabin     Embarked  
0       0         A/5 21171     7.25  <NA>  Southampton  
1       0          PC 17599  71.2833   C85    Cherbourg  
2       0  STON/O2. 3101282    7.925  <NA>  Southampton  
3       0            113803     53.1  C123  Southampton  
4       0            373450     8.05  <NA>  Southampton  
..    ...               ...      ...   ...          ...  
886     0            211536     13.0  <NA>  Southampton  
887     0            112053     30.0   B42  Southampton  
888     2        W./C. 6607    23.45  <NA>  Southampton  
889     0            111369     30.0  C148    Cherbourg  
890     0            370376     7.75  <NA>   Queenstown  

[891 rows x 12 columns]

In [76]:
passengers.loc[:,'Survived'].value_counts()# Looking for survive status

False    549
True     342
Name: Survived, dtype: int64

In [88]:
age_survival=passengers.loc[:,['Survived','Age']].groupby('Survived').mean()
age_survival

Age
Survived           
False     30.626179
True       28.34369

In [89]:
median_age_survival=passengers.loc[:,['Survived','Age']].groupby('Survived').median()
median_age_survival

Age
Survived      
False     28.0
True      28.0

In [ ]:
passengers.iloc[0:5,[2,3,-1]] # first 5 passangers, only columns 2, 4 and the last one

In [ ]:
passengers.loc[10:20,['Age', 'Fare']] 
# passengers with indexes 10 to 20 passangers, only columns 'Age', 'Fare'

In [ ]:
passengers.loc[:,['Name']] # All rows, column name

In [ ]:
# finally, you can combine them:
# get first 5 people, and their Names
passengers.iloc[0:5,:].loc[:,'Name']

To pick first or last n items you can also use .head(n) or tail(n).

And if picking just one column (and it does not have spaces in the name), you can use .ColumnName 

In [ ]:
passengers.head(5).Name

In [ ]:
# notice that .loc[:,'Name'] is the same as .Name
# try to guess, what are advantages and limitations of .ColumnName notation? 

passengers.head(5).loc[:,'Name']

Ok, let;s find our 5 questions and answer them (call them A,B,C,D and E)

### A: Names of oldest passangers. What can we tell about their Survival rate when compared with youngest ones?

here's a start on the first one

In [ ]:
# just select name and age components of the data frame. 
dfA2 = passengers[['Age', 'Name', 'Pclass', 'Survived']].copy()

# Sort them by Age 
dfA2.sort_values(by='Age', ascending=False, inplace=True)

# and get the names of 10 oldest passangers
oldest_10 = dfA2.iloc[0:10]
oldest_10

In [ ]:
# note: and if you wanted to get names of all people over 70
print(dfA2[dfA2.Age > 70].loc[:, ['Name','Age']])

In [ ]:
# Or get names of all people over 50 who survived
print(dfA2[(dfA2.Age > 50)&(dfA2.Survived == 1)].loc[:, ['Name','Age']])

In [ ]:
print('oldest 10 survival', dfA2.iloc[0:10,:].loc[:, 'Survived'].mean())
# could also be
print('oldest 10 survival', dfA2.iloc[0:10,:].Survived.mean())
print('oldest 10 survival', dfA2.head(10).Survived.mean())

In [ ]:
print('youngest 10 survival', dfA2.tail(10).Survived.mean())

In [ ]:
# wait a minute.. that looks a bit suspicious! children died with such frequency?
# let's have a look at the data!
dfA2.tail(10)

In [ ]:
# oh no! that's the data for those where age is missing!
# let's ask for average for those with know age, but still youngest
dfA2[ dfA2['Age'].notna()].tail(10)
# why do you think decimal ages are 0.42, 0.67, 0.75, 0.83 and 0.93 - oh why?

In [ ]:
print("youngest 10 survival: ", dfA2[ dfA2['Age'].notna()].tail(10).Survived.mean())


In [ ]:
print('average age survival', dfA2.Survived.mean())
print('average age survival (if age known)', dfA2[ dfA2['Age'].notna()].Survived.mean())
# how do you interpret those results? What would it mean?

Answer: Name of the oldest passanger (80) was Mr. Algernon Henry Wilson Barkworth. Survival rate of 10 oldest passangers was 10%, and of the youngest 10 passagners 90%. In context, the average survival rate was 38%.

### B: What's the impact of Passanger Class and Sex on the suvival?

In [ ]:
# your code here (add more cells with '+' in the top menu)

You answer here:

### C: Your question? Select a subset, and look at it's qualities

In [ ]:
# your code here (add more cells with '+' in the top menu)

You answer here:

### D: Your question?

In [ ]:
# your code here (add more cells with '+' in the top menu)

You answer here:

### E: Your question?

In [ ]:
# your code here (add more cells with '+' in the top menu)

You answer here:

# Task 3: Cleaning up and improving the dataset:

A: Identify missing values and decide what to do with them. You can remove whole rows with missing things, or you can fill them with averages. Or averages for a particular subgroup.

B: Rename the column names so that they follow the same naming convention (eg. no spaces, CamelCase or snake_case). 

C: Can you create new columns which use available information? eg. Can you create a column TravelingAlone based on the values in siblings/partners and parents/children ?

### A: Looking at missing values

In [ ]:
# btw. re-run this cell to reset yoru dataset it its original form
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')


In [ ]:
# all data into binary: is missing or not:
passengers.isna()

In [90]:
# by columns, are any items True?
passengers.isna().any()

PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
Age             True
SibSp          False
Parch          False
Ticket         False
Fare           False
Cabin           True
Embarked        True
dtype: bool

In [10]:
# how many passagners would you loose if you passengers.dropna(axis=0) ?
# you can use .shape()
pass_no_na = passengers.dropna(axis=0)
pass_no_na

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803     53.1         C123        S  
6        0     17463  51.8625          E46        S  
10       1   PP 9549     16.7           G6        S  
11       0    113783    26.55         C103        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695      5.0  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053     30.0          B42        S  
889      0    111369     30.0         C148        C  

[183 rows x 12 columns]

In [11]:
pass_no_na=pass_no_na.loc[:,['Survived','Age']].groupby('Survived').median()
pass_no_na

Age
Survived      
0         44.5
1         33.0

In [ ]:
# You could replace all missing values with a set value, 
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
pass_no_na = passengers.fillna(0) 
# but what unexpected effects it could have?
pass_no_na

In [ ]:
# or you can replace them with a most common value,
# or other statistic of that column eg.mean, max
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
passengers.fillna(passengers.mean(numeric_only=True)) 

In [ ]:
# for non-numeric values one of the oprtions is to use the most common (mode) value.
# since there might be mady most common values, you'll have to specify that you want the first one
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
passengers.fillna(passengers.mode().iloc[0]) 
# notice what happens with the Cabin

In [12]:
# you could also sopecify individual columns and values
#passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
passengers.Cabin = passengers.Cabin.fillna("NoCabin") 
passengers

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  <NA>      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare    Cabin Embarked  
0        0         A/5 21171     7.25  NoCabin        S  
1        0          PC 17599  71.2833      C85        C  
2        0  STON/O2. 3101282    7.925  NoCabin        S  
3        0            113803     53.1     C123        S  
4        0            373450     8.05  NoCabin        S  
..     ...               ...      ...      ...      ...  
886      0            211536     13.0  NoCabin        S  
887      0            112053     30.0      B42        S  
888      2        W./C. 6607    23.45  NoCabin        S  
889      0            111369     30.0     C148        C  
890      0            370376     7.75  NoCabin        Q  

[891 rows x 12 columns]

In [ ]:
# Your task, decide what to do with the missing values in the origibnal dataset:
# Once your_dataset.isna().any() returns all False, there are no more missing values

In [13]:
#passengers = passengers.dropna(subset = ['Age'])
passengers.Age = passengers.Age.fillna(np.ceil(passengers.Age.mean()))
passengers

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  30.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare    Cabin Embarked  
0        0         A/5 21171     7.25  NoCabin        S  
1        0          PC 17599  71.2833      C85        C  
2        0  STON/O2. 3101282    7.925  NoCabin        S  
3        0            113803     53.1     C123        S  
4        0            373450     8.05  NoCabin        S  
..     ...               ...      ...      ...      ...  
886      0            211536     13.0  NoCabin        S  
887      0            112053     30.0      B42        S  
888      2        W./C. 6607    23.45  NoCabin        S  
889      0            111369     30.0     C148        C  
890      0            370376     7.75  NoCabin        Q  

[891 rows x 12 columns]

### B: Duplicates. 

In this dataset there are no dulplicates. So here's a small exercise. Let's duplicate the first person so they appear 3 times. And let's change the age of one of those entries. So we're pretending first passanger got accidentally added 3 times, including once with a slightly different data for age.

In [ ]:
# att two more copies of first row
passengers = pd.read_csv('./data/titanic_kaggle_train.csv')
with_duplicates = passengers
with_duplicates = pd.concat((passengers.iloc[0:1], passengers.iloc[0:1], passengers))
with_duplicates

In [ ]:
# now let's fix the index
with_duplicates.reset_index(drop=True, inplace=True) # drop completely forgets previous index
with_duplicates

In [ ]:
# change one of the duplicates to have another age:
with_duplicates.loc[1,'Age'] = 23 

# Note: as well as loc[] and iloc[], which access many items,
# you can use at[] and iat[] to access just one:
# with_duplicates.at[1,'Age'] = 23
with_duplicates

Now let's check for duplicates, and remove them. You might want to define a duplicate as when some unique element (or a combination of unique elements) are the same.

In [ ]:
with_duplicates.duplicated()
# here you should see that the person at index 2 is a duplicate (perfectly identical)
# you could also decide to keep last (not first) with  duplicated(keep='last')

In [ ]:
no_duplicates1 = with_duplicates.drop_duplicates()
no_duplicates1
# this would remove all completely-dentical duplicates

In [ ]:
# this would remove all duplicates, identical only in those columns
no_duplicates2 = with_duplicates.drop_duplicates(subset=['Name'])
no_duplicates2
# notice we're down to one duplicated first passanger

In [ ]:
# so now there are no duplicates
no_duplicates2.duplicated()

In [ ]:
# and we can re-index
no_duplicates2.reset_index(drop=True, inplace=True)
no_duplicates2

### C: Improving Column names

Look at the notes if you're unsure how to do it. You're looking for rename(). Aim for no spaces or special characters in column names, so that you can use . notation, like .Age

One improvementc could be making those columns more human readable: SibSp, Parch. See link to Kaggle above for more insights or what they mean/

In [95]:
passengers = passengers.rename(columns = {'Pclass': 'Passenger_Class', 'SibSp': 'Siblings_Spouses', 'Parch': 'Parents_Children'})
passengers

PassengerId  Survived Passenger_Class  \
0              1     False             3rd   
1              2      True             1st   
2              3      True             3rd   
3              4      True             1st   
4              5     False             3rd   
..           ...       ...             ...   
886          887     False             2nd   
887          888      True             1st   
888          889     False             3rd   
889          890      True             1st   
890          891     False             3rd   

                                                  Name     Sex   Age  \
0                              Braund, Mr. Owen Harris    male  22.0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
2                               Heikkinen, Miss. Laina  female  26.0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
4                             Allen, Mr. William Henry    male  35.0   
..                                                 ...     ...   ...   
886                              Montvila, Rev. Juozas    male  27.0   
887                       Graham, Miss. Margaret Edith  female  19.0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  <NA>   
889                              Behr, Mr. Karl Howell    male  26.0   
890                                Dooley, Mr. Patrick    male  32.0   

    Siblings_Spouses Parents_Children            Ticket     Fare Cabin  \
0                  1                0         A/5 21171     7.25  <NA>   
1                  1                0          PC 17599  71.2833   C85   
2                  0                0  STON/O2. 3101282    7.925  <NA>   
3                  1                0            113803     53.1  C123   
4                  0                0            373450     8.05  <NA>   
..               ...              ...               ...      ...   ...   
886                0                0            211536     13.0  <NA>   
887                0                0            112053     30.0   B42   
888                1                2        W./C. 6607    23.45  <NA>   
889                0                0            111369     30.0  C148   
890                0                0            370376     7.75  <NA>   

        Embarked  
0    Southampton  
1      Cherbourg  
2    Southampton  
3    Southampton  
4    Southampton  
..           ...  
886  Southampton  
887  Southampton  
888  Southampton  
889    Cherbourg  
890   Queenstown  

[891 rows x 12 columns]

### D: Categorical variables (in R called Vectors, in Java called Enums)

Some values are presented as numbers (Pclass), but really, they are categories. Some are presented as strings (sex), but with some careful consideration and kind design can be analysed as categories too (notice this very old dataset is using Sex, not gender*).

**Note on usage of Gender and Sex in datsets: read up on this after the lab**

*In all modern datasets Gender is represented as much more complex datastructure than just True/False binary (this was common in early days of databases, where gender was a True/False value 'is_male'). Later it was a category (with 2 or 3 choices eg. 'male', 'female', 'prefer_not_to_say'). 

Luckilly, to reflect beauty of human experience there are some very interesting research projects on how to design inclusive forms and datsets. This Scottish Government oriject involved data scientists and people with various experiences of gender. Do some reading on the subject https://www.gov.scot/publications/reviewing-design-methods-make-more-sensitive-gender/pages/1/ and notice it does not provide 'one answer fits all' but rather encourages empathy on project-by-project basis. Also form designers together with LGBTQI+ organisations are constanly advancing this field https://uxdesign.cc/designing-forms-for-gender-diversity-and-inclusion-d8194cf1f51  

So let's not look at the Sex field, but rather Pclass.

Let's first have a look at exploring Passanger class: What passanger classess were there on the titanic (on british trains we still have 1st and 2nd class carriages

In the dataset they are represented 

In [ ]:
passengers.Pclass

In [ ]:
passengers.Pclass.value_counts()

In [ ]:
# are there any NA ?
passengers.Pclass.isna().value_counts()

In [ ]:
passengers['TicketClass'] = pd.Categorical(passengers.Pclass)
passengers

In [ ]:
# Categorical data has number of advantages in generating statistics etc
# but also make our datasts much smaller (in memory) and faster (irn processing)

passengers.memory_usage()
# notice ticketclass stores the same data as Pclass, but uses 15% of memory.

In [ ]:
# Additionally you can specify the order of categories, so that you 
# could perform comparisons

# introduce order
passengers['TicketClass'] = pd.Categorical(passengers.Pclass, 
                                           categories=[3,2,1], # here specify order
                                           ordered=True
                                          )
# optionally re-name categories with cat.rename_categories
mapping = {1: 'Luxury', 2: 'Economy', 3: 'LowerDeck'}
passengers['TicketClass'] = passengers['TicketClass'].cat.rename_categories(mapping)
passengers['TicketClass'] # notice printout shows order of categories

In [ ]:
# so that you can now eg choose all TicketClass LOWER THAN a number:
passengers[passengers['TicketClass'] < 'Luxury']

In [ ]:
# what other columns could be categories. Try them!

### E: Creating more columns

Look at the notes if you're unsure how to do it. What could you use: age groups? In one og the notebooks you were shown how to group different ranges of values into groups. Maybe you could use that technique on age?

Or maybe you could combine columns abotu partners and dependants (SibSp,Parch) into a column TravelsAlone?

How do Survival stats look like for different values of the new columns?